Local Selenium Grid

glogiotatidis opened the pull request:
Second commit will be removed before merging, only there for testing

Test run https://gitlab.com/mozmeao/www-config/-/pipelines/204608823

## Details
 - Firefox, Chrome runs on runners
 - Replaced xdist with parallel. See https://github.com/mozilla/bedrock/pull/9565
 - Failed tests upload Screenshots and logs on Gitab, accessible via the IU
![image](https://user-images.githubusercontent.com/584352/96456488-3a88cb00-1227-11eb-96f0-b69190effabb.png)
 - Tests take about ~16-20 minutes to complete (on a `c5a` runner, see https://github.com/mozmeao/infra/issues/1359)

This is still in progress but I'd like your feedback @alexgibson @pmac. It's missing re-running failed tests on Saucelabs to get videos.

alexgibson reviewed (commented):
This is looking great, nice work! A couple of notes / observations:

- I think we can remove the Edge test job. We're only running this against Dev right now as a test, and in terms of coverage it likely overlaps Chrome now that it's moved over to Chromium.
- We can also disable download tests for dev deployments, as per https://github.com/mozmeao/www-config/pull/287
- Screenshots for failed tests will definitely be helpful 🦾 
- Chrome tests took ~6m, compared to ~18m with SauceLabs ⚡ 
- Firefox tests took ~12m, compared to ~25m with SauceLabs ⚡

glogiotatidis commented:
Features:
 - Simplified the pipeline, for both `deploy` and `test` stages
 - Added manual jobs for firefox, chrome, download and edge on Saucelabs. I.e. One can clicks the "Play" button on gitlab to get them running against saucelabs when / if needed. E.g. when videos of failures are needed. Edge is still available but only triggered manually
 - Firefox, Chrome, Download, Headless tests run on our runners
 - We store artifacts (screenshots and pytest-html output) for failed jobs on gitlab. Available for browse and download via its UI
 - Download tests run only when we deploy to environments other than dev

Latest test run https://gitlab.com/mozmeao/www-config/-/pipelines/206202775

alexgibson commented:
This is all looking great to me @glogiotatidis - is there anything outstanding that still needs doing before mergin this? 

One thing I'm not quite clear on is if this will still run the full suite of tests when pushing to `run-integration-tests`? Other than that, everything seems to be in order.

alexgibson commented:
Latest test run https://gitlab.com/mozmeao/www-config/-/pipelines/206202775

Hmm, I notice here the firefox and chrome remote tests are taking ~20mins. That seems significantly longer than previous runs?

glogiotatidis commented:
> Hmm, I notice here the firefox and chrome remote tests are taking ~20mins. That seems significantly longer than previous runs?
> 
> Edit: looks like they are running using only 2 processes?

We're still running on the old runners. We'll upgrade a next days: https://github.com/mozmeao/infra/issues/1359

glogiotatidis commented:
> One thing I'm not quite clear on is if this will still run the full suite of tests when pushing to `run-integration-tests`? Other than that, everything seems to be in order.

I don't see any references to `run-integration-tests` in the original `.gitlab-ci`? Can you explain what's supposed to do?

alexgibson commented:
I'm not 100% sure how all the plumbing works, but the process works as follows:

- Push a commit to mozilla/bedrock on the run-integration-tests branch
- That triggers [this job](https://github.com/mozmeao/www-config/blob/master/.gitlab-ci.yml#L66-L81), which takes the commit, deploys it to a demo, and then runs the full test suite against it.

This is something the front-end team use quite a bit to help test large changes before merging. If you want to chat about this more, let me know!

glogiotatidis commented:
Got it. I removed `oregon-b` references b/c we don't have this cluster anymore. @pmac maybe you know how this is supposed to work now?

pmac commented:
I forgot that the `bedrock-test` app was still in oregon-b. So I guess that ability was collateral damage from the oregon-b implosion. We'll have to setup a new app for doing those test runs in GCP.

pmac commented on .gitlab-ci.yml:
```diff
@@ -2,38 +2,35 @@ stages:
+    when:
+      on_success
+    expire_in:
+      15 mins
```
Why keep an artifact for only 15 min?

alexgibson replied:
If this relates to a screenshot or html report, it would be useful to keep those around at least for a day or so.

pmac replied:
this is different. this is an artifact for when the deployment succeeds.

glogiotatidis replied:
This artifact is used to pass `BASE_URL` to the `test` stage jobs. It's not really useful to keep longer. The output from the tests are kept for a week

alexgibson commented:
> I forgot that the bedrock-test app was still in oregon-b. So I guess that ability was collateral damage from the oregon-b implosion. We'll have to setup a new app for doing those test runs in GCP.

Got it. In that case, would it be best to move forward and merge this change now, and then we can fix the bedrock-test app in a followup?

pmac reviewed (commented):
This is looking great. One thing I don't understand yet is how the tests know which `BASE_URL` to use. Does that variable carry over from the deploy phase?

alexgibson reviewed (approved):
Needs a rebase but from my perspective this is all looking great. Thanks for all your work @glogiotatidis 🌮
